Adding validation step for missing required ClusterTemplate's cluster…

edcdavid opened the pull request:
Adding validation step for missing required ClusterTemplate's clusterInstanceParameters fields:
- nodeClusterName
- oCloudSiteId
- policyTemplateParameters
- clusterInstanceParameters
Testing the for the presence of the required schema keys in the properties and in the required sections

edcdavid commented:
/cc @Missxiaoguo @irinamihai

irinamihai commented on internal/controllers/clustertemplate_controller.go:
```diff
@@ -327,6 +334,54 @@ func generateTemplateID(ctx context.Context, c client.Client, object *provisioni
+// - policyTemplateParameters
+// - clusterInstanceParameters
+func validateTemplateParameterSchema(object *provisioningv1alpha1.ClusterTemplate) error {
+	mandatoryParams := []string{"nodeClusterName",
```
All these values will probably be standardized, so they can be made constants.

Missxiaoguo replied:
There are existing constants for these values.

edcdavid replied:
Fixed

edcdavid replied:
Fixed

irinamihai commented on internal/controllers/clustertemplate_controller.go:
```diff
@@ -327,6 +334,54 @@ func generateTemplateID(ctx context.Context, c client.Client, object *provisioni
+	var missingParameter []string
+	for _, param := range mandatoryParams {
+		_, err := utils.ExtractSubSchema(object.Spec.TemplateParameterSchema.Raw, param)
+		if err != nil && strings.HasPrefix(err.Error(), fmt.Sprintf("subSchema %s does not exist:", param)) {
```
Maybe refactor these lines to smth like:
```
if err != nil {
    if strings.HasPrefix(err.Error(), fmt.Sprintf("subSchema %s does not exist:", param)) {
        missingParameter = append(missingParameter, param)
    } else {
        return fmt.Errorf("error extracting subschema at key %s: %w", param, err)
    }
}
```

edcdavid replied:
Fixed

irinamihai commented on internal/controllers/clustertemplate_controller.go:
```diff
@@ -327,6 +334,54 @@ func generateTemplateID(ctx context.Context, c client.Client, object *provisioni
+			missingRequired = append(missingRequired, param)
+		}
+	}
+	validationFailureReason := fmt.Sprintf("failed to validate ClusterTemplate name:%s. ", object.Name)
```
I think we can remove `name` from this

edcdavid replied:
Fixed

irinamihai commented on internal/controllers/clustertemplate_controller.go:
```diff
@@ -327,6 +334,54 @@ func generateTemplateID(ctx context.Context, c client.Client, object *provisioni
+		validationFailureReason = fmt.Sprintf(" The following mandatory fields are missing: %s", strings.Join(missingParameter, ","))
+	}
+	if len(missingRequired) != 0 {
+		if len(missingParameter) != 0 {
```
If you end the reason above with a `.` ()line 372 and start the one below with a `The`, you don't need this check.

edcdavid replied:
Fixed.

Missxiaoguo reviewed (commented):
Could we also add the validation for the types? for example, 
`nodeClusterName` and `oCloudSiteId` should be of "type: string" and `policyTemplateParameters` and `clusterInstanceParameters` should be of "type: object". That's the key for validating the input types against the schema.

edcdavid commented:
> Could we also add the validation for the types? for example, `nodeClusterName` and `oCloudSiteId` should be of "type: string" and `policyTemplateParameters` and `clusterInstanceParameters` should be of "type: object". That's the key for validating the input types against the schema.

Fixed. Also added unit tests.

Missxiaoguo reviewed (commented):
/lgtm

irinamihai commented:
/approve
